### ISSUE TITLE
False positive 'Expression expected' error in async generator

### ISSUE THREAD

**Issue by ajafff** [microsoft/TypeScript#24090]
**TypeScript Version:**  2.9.0-dev.20180512

**Search Terms:**  await yield

**Code**

```ts
async function* test() {
    await (<any>yield); // error: Expression expected
}
```

**Expected behavior:**

Parses correctly and is transpiled to `await (yield);`.
Maybe this should also work for `await <any>yield;`, which should also transpile to `await (yield);`

It already works for `await (yield);` so why is it different when adding a type assertion?

**Actual behavior:**

Error: Expression expected.

**Playground Link:** https://agentcooper.github.io/typescript-play/?target=5#code/IYZwngdgxgBAZgV2gFwJYHsICobIKYjIAUAlDAN4BQMMwA7sKsjEQDzARgB8YqeANgBMSAbmq0GTGO048+QsTXqNmvAYMUSVLNUNHjlUmdyK7hYgL5A

**Related Issues:**

**Comment by Kingwl** [comment]
yield is not the UnaryExpression and await is only leading UnaryExpression

does ECMA spec support this? (seems disallow, maybe i'm wrong )

**Comment by RyanCavanaugh** [comment]
I think this is actually working as intended. Parentheses containing only type assertions are removed since there are times when you *can't* have parens but *must* have an assertion, so e.g.
```ts
async function* test() {
    await (<any>foo);
}
```
turns into
```ts
async function* test() {
    await foo;
}
```

So the correct thing to write here is
```ts
async function* test() {
    await (<any>(yield));
}
```
which parses as expected
